### ISSUE TITLE
Feature suggestion for multilingual sites in joomla

### ISSUE THREAD

**Issue by hjelperne** [thexmanxyz/Cookie-Consent-Neo-Gantry#5]
Hi,

Nice cookie consent!
I was wondering if it would be possible to add a feature for multi-lingual sites (in joomla).

Let me explain:
Not all gantry sites use different template outlines for different languages (although that is an option).
In the case where the same Base outline (for instance) is used for all languages it would be extremely nice to have the option to select language (for language specific text) for the Cookie consent Neo Atom.

**Suggestion  1 (preferred from my perspective):**

The option to add for instance two cookie consent atoms to the Base Outline with different languages selected in an option field (Defaults to: All).
One cookie consent atom for f.ex English and another for f.ex Swedish.
Then the correct one will get pulled according to the selected/default language on the site.
This way translations are maintained in one location and there is no need to create yet another (or series of) Outlines just to accommodate translation efforts of the text in the cookie consent.

**Suggestion 2 (alternative approach):**

A second option could of course be to add "translation overrides" possibilities to the text fields.
So one can provide strings from the "Extensions -> Languages -> Overrides" in the text fields instead of the actual text. An option "use translation overrides" or similar would have to present.

Hope the suggestions are appreciated.

Anyway: Great stuff!

Hjelperne

**Comment by thexmanxyz** [comment]
@hjelperne Hi, thanks for your words and for your feature request.

First of all, currently the Atom works for all Gantry supported CMS and I want to maintain that compatibility. So we are probably a bit limited on how to implement the multi language behavior you described. I like the idea and I also think that it would be very useful if we can integrate this feature somehow. Moreover, I understand that it might be not very clean if you need one Outline per language. That would be hard to maintain and is also not really a great way of solving the issue just for one Atom.

1.) I understand what you have written but I don't know a method how to tell Gantry which Atom should be loaded if the Atom is defined multiple times in dependence of a condition. Maybe you can clarify how this should work because I currently do NOT know a way to integrate this functionality.

2.) Generally, this seems to be a Joomla only approach. With the option "use translation overrides" it would be possible to maintain compatibility with Grav and Wordpress however the multilanguage support would be limited to Joomla. Moreover IDK how to fetch the language overrides within the Atom code. I don't know if this is even possible. Do you have a snippet or reference how to fetch that override information?

Alternative: One thing I have thought that would be most applicable (if possible) depends if there is a possiblity to fetch the curren CMS language over Gantry. For Joomla this is possible I checked that so you can by means of the determined Language use an alternate output for the Atom (if defined). If not defined the default one is taken.

The suggestions are highly appreciated but I don't know if there is a general solution that even works.

Thanks and looking forward for your thoughts.

**Comment by thexmanxyz** [comment]
Another thing to mention. For particles it is easy to integrate multilanguage support because you can create a particle module. Create for each language a module, set it on a specific position and define the corresponding language. Sadly such a possiblity doesn't exist for Atoms...

**Comment by hjelperne** [comment]
Too true. 
Sorry for my perhaps unenlightened question - but why isnt this a particle (as it contains configurable text etc) in the first place..?

**Comment by thexmanxyz** [comment]
@hjelperne Because it doesn't print any visible HTML direclty like a Particle would do. The only modification that is done on the DOM are the `<script>` and `<link>` tags. So it was created as an Atom because it doesn't print anything in an section it would be attached to.

**Comment by thexmanxyz** [comment]
@hjelperne I think that this is the use case of an Atom. However as we found out there might be other things involved like internationalization which may require a paradigm switch. But I still wait on an answer of you so that we can further evaluate a proper solution.

**Comment by hjelperne** [comment]
I just came across: "Gantry Content Finders" (available since gantry 5.2.14): 
https://github.com/gantry/docs/blob/develop/pages/01.gantry5/05.advanced/10.content-in-particles/docs.md
OR:
http://docs.gantry.org/gantry5/advanced/content-in-particles

_Perhaps it would be possible to replace the language code with a dynamic variable (current lang):
{% do finder.language(**'en-GB'**) %}
And somehow access the translations in a non CMS-specific manner (from default translation ini files / customisable through translation overrides in joomla context) similar to :
{% set options = {key: 'neo.cookie.consent.mytext'} %}
{% set finder = joomla.finder('translation', options) %}

**This is just a WILD SHOT**! Dont know if anything similar to the non-existing pseudo-code above would work at all.
Hoping that you or someone else is more susceptible to actually understand and elaborate on the documentation for _gantry content finders_ to ultimately  produce implementable code.

The idea would be that the extension installs the default translation ini files (some) to the CMS system (and manually added and updated ini files would get picked up if available - if not default to default lang)_ 

**OR maybe the |trans filter is sufficient?
http://docs.gantry.org/gantry5/advanced/twig-filters**

**Comment by thexmanxyz** [comment]
@hjelperne Hey there sorry for the late answer but I had to finally ask at the Gantry repository / community for a solution and we finally got one. Thanks for your efforts. Just short on your message

The content finder does not work because it is only for retrieving articles and we can't extract language information from there (not in a meaningful way). Especially not the current active language. As I already said we need here a generic solution because we have to rely on three CMS (Wordpress, Joomla and Grav) so we can't take Joomla only functionality e.g. language overrides or language files. However, don't mind further the link below shows how I can retrieve the current language and I will use that information within the next revision and update the Atom appropriately:

[#2287](https://github.com/gantry/gantry5/issues/2287)

As an additonal information, I will create a list were you can add lang codes for all languages you need. And the one active on the site will be taken to render the GUI elements. So you only need one Atom where you specify all language strings for every language you have on your site for Cookie Consent. That is also my preferable solution.

Ad.: It would be possible to create ini files (if Joomla is the CMS) for the particle and translate them with `|trans` filter however I won't do that because it would over complicate the manual installation. Because for Grav and Wordpress we would need separate install methods / instructions. That's something I maybe consider for a release somewhere in the future.

**Comment by hjelperne** [comment]
Hi again,
Thats great news. Good finding!
Do you have any idea when the feature might be available?
Thanks!

**Comment by thexmanxyz** [comment]
@hjelperne I try to finish it in the next days. Shouldn't be that much of changes. I keep you updated.

**Comment by thexmanxyz** [comment]
@hjelperne Fixed with https://github.com/thexmanxyz/Cookie-Consent-Neo-Gantry/commit/b3b1dce0a4ed1a000c096eec5483c4e1c5941439. Latest version which includes the new functionality available now: [v1.5.0](https://github.com/thexmanxyz/Cookie-Consent-Neo-Gantry/releases/tag/v1.5.0)
